File: Tree/Binary Tree/Problem Solving/Leetcode/Most Frequent Subtree Sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int frequentTreeSum(TreeNode* root, map<int,int> &m, int &maxFrequency){
        if(root == nullptr)return 0;

        int left = frequentTreeSum(root->left, m, maxFrequency);
        int right = frequentTreeSum(root->right, m, maxFrequency);
    
        int treeSum = root->val+left+right;
        m[treeSum]++;
        maxFrequency = max(maxFrequency,m[treeSum]);

        return treeSum;
    }
    vector<int> findFrequentTreeSum(TreeNode* root) {
        vector<int>v;
        map<int,int>m;
        int maxFrequency = 0;
        frequentTreeSum(root,m,maxFrequency);
        
        for(auto const& pair : m){
            if (maxFrequency == pair.second){
                v.push_back(pair.first);
            }
        }
        return v;
    }
};

File: Tree/Binary Tree/Problem Solving/Leetcode/Most Frequent Subtree Sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def frequentTreeSum(self, root=None)->int:
        if not root:
            return 0

        left =  self.frequentTreeSum(root.left)
        right = self.frequentTreeSum(root.right)
        
        treeSum = root.val + left + right
        self.freq_dict[treeSum] += 1
        self.max_freq = max(self.max_freq,self.freq_dict[treeSum])
        return treeSum
        
    def findFrequentTreeSum(self, root: Optional[TreeNode]) -> List[int]:
        self.freq_dict = collections.defaultdict(int)
        self.max_freq = 0
        self.ans = []
        self.frequentTreeSum(root)
        for key,value in self.freq_dict.items():
            if value == self.max_freq:
                self.ans.append(key)
        return self.ans

